### PR TITLE
[SPARK-33084][TEST][FollowUP] Fix a flaky test in SparkContextSuite

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

The test case `SPARK-33084: Add jar support Ivy URI -- transitive=true will download dependency jars` in `SparkContextSuite` is becoming flaky:

- https://github.com/gengliangwang/spark/runs/4698825652?check_suite_focus=true
- https://github.com/gengliangwang/spark/runs/4698331067?check_suite_focus=true
- https://github.com/AngersZhuuuu/spark/runs/4697626841?check_suite_focus=true

The reason is that some of the events in `LogAppender` are null so that there is NPE:
```
[info]   Cause: java.lang.NullPointerException:
[info]   at org.apache.spark.SparkContextSuite.$anonfun$new$128(SparkContextSuite.scala:1077)
[info]   at org.apache.spark.SparkContextSuite.$anonfun$new$128$adapted(SparkContextSuite.scala:1077)
```
This PR is to fix the issue to unblock PR builders.  

### Why are the changes needed?

Fix flaky test

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Just tests

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @AngersZhuuuu as well

**Review by HyukjinKwon — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/SparkFunSuite.scala [hunk 1]
```diff
@@ -284,5 +284,7 @@ abstract class SparkFunSuite
     def setThreshold(threshold: Level): Unit = {
       _threshold = threshold
     }
+
+    def nonNullLogEvents(): Seq[LogEvent] = loggingEvents.filterNot(_ == null)
```
Comment (gengliangwang) [inline core/src/test/scala/org/apache/spark/SparkFunSuite.scala:288]: Actually,  I tried overwriting `append(loggingEvent: LogEvent)` to filter out the null events, but it doesn't work. I don't know why there are null events yet but I think we should fix this to unblock the CI tests.

**Review by AngersZhuuuu — APPROVED** [review]
LGTM,  but I want to know, is this issue happen after we upgrade log4j2?

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by gengliangwang** [comment]
> LGTM, but I want to know, is this issue happen after we upgrade log4j2?

I am not sure either. cc @viirya

**Review by viirya — APPROVED** [review]

**Review by viirya — COMMENTED** [review]
I've addressed this issue after the migration by making stored log event as immutable. When I looked at the issue locally, I found in `append` the `LogEvent` isn't null. But log4j2 seems to release `LogEvent` objects later. I did adding `toImmutable` here and it reduced the flakyness as I don't see it again in later PRs so I thought it was fixed at all.

Although it is rarely now, but seems it is still happened.

Maybe more reliable way is to store the message directly in `append`.

**Comment by gengliangwang** [comment]
@HyukjinKwon @dongjoon-hyun @AngersZhuuuu @viirya Thanks for the review!
Merging to master

**Pull request closed without merging**